services/wallet/async/async.go

```diff
@@ -155,3 +160,57 @@ func (d *AtomicGroup) Error() error {
 func (d *AtomicGroup) Stop() {
 	d.cancel()
 }
+
+func NewQueuedAtomicGroup(parent context.Context, limit int64) *QueuedAtomicGroup {
+	return &QueuedAtomicGroup{NewAtomicGroup(parent), limit, 0, []Command{}}
+}
+
+type QueuedAtomicGroup struct {
+	*AtomicGroup
+	limit       int64
+	count       int64
+	pendingCmds []Command
+}
+
+func (d *QueuedAtomicGroup) Add(cmd Command) {
+	counter := atomic.LoadInt64(&d.count)
```

Comment:
Hmm it seems to me that we're making the counter thread-safe by using atomic ops, but not the list of commands, which might cause trouble (for example, two threads adding commands while `counter >= d.limit` will try to append to the list at the same time).